src/libNode/Node.cpp

```diff
@@ -271,43 +228,257 @@ void Node::Prepare(bool runInitializeGenesisBlocks) {
+             cv_status::timeout);
+  }
+
+  if (m_mediator.m_txBlockChain.GetBlockCount() > oldTxNum + 1) {
+    LOG_GENERAL(INFO, "Node recovery too late, apply re-join process instead");
+    return false;
+  }
 
-  tDS.join();
-  tTx.join();
+  /// Retrieve lacked final-block state-delta from lookup nodes
+  if (m_mediator.m_txBlockChain.GetBlockCount() > oldTxNum) {
+    unique_lock<mutex> lock(
+        m_mediator.m_lookup->m_MutexCVSetStateDeltaFromSeed);
+
+    do {
+      m_mediator.m_lookup->GetStateDeltaFromLookupNodes(
+          m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum());
+      LOG_GENERAL(
+          INFO,
+          "Retrieve final block state delta from lookup node, please wait...");
+    } while (m_mediator.m_lookup->cv_setStateDeltaFromSeed.wait_for(
+                 lock, chrono::seconds(RECOVERY_SYNC_TIMEOUT)) ==
+             cv_status::timeout);
+  }
+
+  /// Save coin base for final block, from last DS epoch to current TX epoch
+  for (uint64_t blockNum =
+           m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() -
+           (m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() %
+            NUM_FINAL_BLOCK_PER_POW);
+       blockNum <=
+       m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum();
+       ++blockNum) {
+    LOG_GENERAL(INFO, "Update coin base for finalblock with blockNum: "
+                          << blockNum << ", reward: "
+                          << m_mediator.m_txBlockChain.GetBlock(blockNum)
+                                 .GetHeader()
+                                 .GetRewards());
+    m_mediator.m_ds->SaveCoinbase(
+        m_mediator.m_txBlockChain.GetBlock(blockNum).GetB1(),
+        m_mediator.m_txBlockChain.GetBlock(blockNum).GetB2(), -1, blockNum + 1);
+    m_mediator.m_ds->m_totalTxnFees +=
+        m_mediator.m_txBlockChain.GetBlock(blockNum).GetHeader().GetRewards();
+  }
+
+  bool ds_result;
+  m_retriever->RetrieveDSBlocks(ds_result, wakeupForUpgrade);
+
+  /// Retrieve sharding structure and setup relative variables
+  BlockStorage::GetBlockStorage().GetShardStructure(
+      m_mediator.m_ds->m_shards, m_mediator.m_node->m_myshardId);
+  LoadShardingStructure();
+  m_mediator.m_ds->ProcessShardingStructure(
+      m_mediator.m_ds->m_shards, m_mediator.m_ds->m_publicKeyToshardIdMap,
+      m_mediator.m_ds->m_mapNodeReputation);
+  m_mediator.m_consensusID =
+      (m_mediator.m_txBlockChain.GetBlockCount()) % NUM_FINAL_BLOCK_PER_POW;
+
+  /// Save coin base for micro block, from last DS epoch to current TX epoch
+  std::list<MicroBlockSharedPtr> microBlocks;
+  if (BlockStorage::GetBlockStorage().GetRangeMicroBlocks(
+          m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() -
+              (m_mediator.m_txBlockChain.GetLastBlock()
+                   .GetHeader()
+                   .GetBlockNum() %
+               NUM_FINAL_BLOCK_PER_POW),
```

Comment:
Same change will be applied here. Btw in 788, MicroBlock will have GetEpochNum rather than GetBlockNum, which is to keep the same convention with ds/vc/fallback block.

Reply:
ok